File: 045_Happy Number.py
"""
Happy Number

Determine if a number is happy: repeatedly replace number with sum of squares of digits;
if it becomes 1 it's happy; detect cycles otherwise.
"""

def next_num(n: int) -> int:
    s = 0
    while n:
        d = n % 10
        s += d*d
        n //= 10
    return s

def is_happy(n: int) -> bool:
    slow = n
    fast = n
    while True:
        slow = next_num(slow)
        fast = next_num(next_num(fast))
        if slow == fast:
            break
    return slow == 1


if __name__ == '__main__':
    print('Happy 1:', is_happy(19))
    print('Happy 2:', is_happy(2))

# Time Complexity: O(log n) per iteration, overall amortized O(1)
# Space Complexity: O(1)

File: 045_Happy Number.cpp
/*
Happy Number

Determine if a number is happy: repeatedly replace number with sum of squares of digits;
if it becomes 1 it's happy; detect cycles otherwise.
*/

#include <bits/stdc++.h>
using namespace std;

int next_num(int n) {
    int s = 0;
    while (n > 0) {
        int d = n % 10;
        s += d*d;
        n /= 10;
    }
    return s;
}

bool isHappy(int n) {
    int slow = n, fast = n;
    do {
        slow = next_num(slow);
        fast = next_num(next_num(fast));
    } while (slow != fast);
    return slow == 1;
}

int main() {
    cout << "Happy 1: " << (isHappy(19)?"true":"false") << endl; // true
    cout << "Happy 2: " << (isHappy(2)?"true":"false") << endl; // false
    return 0;
}

// Time Complexity: O(log n) per iteration, overall amortized O(1) for limited states
// Space Complexity: O(1)
